design/004-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,255 @@
+environment/machine/cluster, it's fairly likely that the data path will be
+incorrect and therefore all recorded workflows will fail.
+
+#### S3
+
+Using S3 in Renku is currently limited to interactive sessions. When launching a
+session, users can specify any S3-compatible URL together with credentials, if
+necessary. The S3 bucket is then mounted in their interactive session.
+Currently, no metadata about these mounts is recorded. The data could, in
+principle, be added to a renku dataset with the "external" mechanism described
+above, but no metadata about its actual origins (i.e. an S3 bucket) would be
+recorded.
+
+#### Upcoming
+
+We have plans to include other external data providers, such as NFS shares,
+openBIS etc. From the user's perspective, they all will generally work in the
+same way as S3; data will be mounted to a local directory outside the repository
+and made accessible as network-mounted storage.
+
+## Design Detail
+
+Generically, we should strive to remove the separation between "hosted" Renku
+(i.e. what I can do if I run an interactive session on renkulab.io) and
+"offline" Renku (what I can do with the CLI anywhere).
+
+Ideally we could once again follow the plug-in paradigm in the CLI, like we
+do for other functionality.
+
+### Provider interface
+
+If we adopt the view that data access should be done on the filesystem level,
+then one approach would be to rely on FUSE as the interface. This means using
+existing FUSE implementations where possible (e.g. S3) or implementing new ones
+using e.g. [`python-fuse` library](https://github.com/libfuse/python-fuse).
+
+
+### Local vs. hosted interactive sessions
+
+Relying on FUSE for providing filesystem access to external resources means that
+locally we do not require special privileges and can manage the data access for
+the user. In the docker-based remote sessions, the situation is more complicated
+because containers require elevated privileges to access FUSE on the host.
+Therefore we will need to expose a microservice that will provide management of
+external data sources for the user. The renku CLI would be aware that it is
+running in a hosted session and instead of executing a local command for
+mounting the data source, it would send off an API request to the service. The
+service itself could run in a sidecar or be integrated with datashim (?), which
+is currently used for providing S3 buckets in interactive sessions. If done via
+a sidecar, we could likely achive dynamic mounting of data sources, which would
+be fantastic for data exploration.
+
+### Commands
+
+#### Adding external data:
+
+The examples below are based on the existing `dataset` commands, but we might
+need to envision something new like `renku data source` which the `dataset`
+commands could reference.
+
+
+```
+renku dataset add s3://server/path/to/bucket
+Access Key: 1234
+Secret Key: abcd
+Store credentials? (y/N): y
+Credentials stored in ~/.renku/renku.ini
```

Comment:
If the credentials are saved in ~/.renku/renku.ini, does this mean that anyone who forks the repo has the credentials?

Reply:
no that's in the user's home directory

Reply:
Oh right. Does that mean the credentials are wiped with each RenkuLab session, i.e. if I enter the credentials in one RenkuLab session, then start a new session a few days later, I have to enter my credentials again?